src/core/connection.cpp

```diff
@@ -33,360 +33,631 @@
-    parser::MessageList* p_messages)
-{
-    assert(p_buffer);
-    assert(p_messages);
 
-    if (!p_websocket_)
-        return websockets::State::ERROR;
+    void Connection::handle_authentication_response(const Message &message)
+    {
+        switch(message.action()) {
+            case Action::ERROR_TOO_MANY_AUTH_ATTEMPTS:
+                {
+                    error_handler_.on_error("too many authentication attempts");
+                    close();
+                } break;
+            case Action::ERROR_INVALID_AUTH_DATA:
+                {
+                    error_handler_.on_error("invalid auth data");
+                    close();
+                } break;
+            case Action::ERROR_INVALID_AUTH_MSG:
+                {
+                    error_handler_.on_error("invalid auth message");
+                    close();
+                } break;
+            case Action::REQUEST:
+                {
+                    assert(message.is_ack());
+                    //TODO: flush message buffer
+                    if (p_login_callback_) {
+                        const auto login_callback = *p_login_callback_;
+                        if (message.num_arguments() < 1) {
+                            login_callback(std::unique_ptr<Buffer>(nullptr));
+                        } else {
+                            const std::unique_ptr<Buffer> auth_data(std::unique_ptr<Buffer>(new Buffer(message[0])));
+                            login_callback(auth_data);
+                        }
+                    }
+                } break;
+            default:
+                {
+                    std::stringstream error_message;
+                    error_message << "authentication message with unknown action" << state_
+                        << ", message header: " << message.header();
+                    error_handler_.on_error(error_message.str());
+                }
+        }
 
-    auto receive_ret = p_websocket_->receive_frame();
-    websockets::State ws_state = receive_ret.first;
-    std::unique_ptr<websockets::Frame> p_frame = std::move(receive_ret.second);
+        ConnectionState new_state = transition_incoming(state_, message);
 
-    if (ws_state == websockets::State::OPEN && !p_frame) {
-        p_buffer->clear();
-        p_messages->clear();
+        if (new_state == ConnectionState::ERROR) {
+            std::stringstream error_message;
+            error_message << "connection error state reached, previous state: " << state_
+                << ", message header: " << message.header();
+            error_handler_.on_error(error_message.str());
+        }
 
-        return ws_state;
+        state(new_state);
     }
 
-    if (ws_state == websockets::State::CLOSED && p_frame) {
-        p_buffer->clear();
-        p_messages->clear();
-
-        close();
+    void Connection::on_error(const std::string &error)
+    {
+        DEBUG_MSG("Websocket error: " << error);
+        state(ConnectionState::ERROR);
```

Comment:
this won't work, as we discussed. An error state is a final state, a websocket error is considered the same as a CLOSE.

You can't recover from a error state, or from a deliberateClose event ( like too many auth attempts or connection/auth timeouts ).